x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvPercentile.java

```diff
@@ -0,0 +1,602 @@
+                    source(),
+                    fieldEval,
+                    percentileEval,
+                    (d) -> new DoubleSortingScratch()
+                );
+                default -> throw EsqlIllegalArgumentException.illegalDataType(field.dataType());
+            };
+            default -> throw EsqlIllegalArgumentException.illegalDataType(field.dataType());
+        };
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new MvPercentile(source(), newChildren.get(0), newChildren.get(1));
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, MvPercentile::new, field, percentile);
+    }
+
+    static class DoubleSortingScratch {
+        private static final double[] EMPTY = new double[0];
+
+        public double[] values = EMPTY;
+    }
+
+    static class IntSortingScratch {
+        private static final int[] EMPTY = new int[0];
+
+        public int[] values = EMPTY;
+    }
+
+    static class LongSortingScratch {
+        private static final long[] EMPTY = new long[0];
+
+        public long[] values = EMPTY;
+    }
+
+    // Evaluators
+
+    @Evaluator(extraName = "DoubleInteger", warnExceptions = IllegalArgumentException.class)
+    static void process(
+        DoubleBlock.Builder builder,
+        int position,
+        DoubleBlock values,
+        int percentile,
+        @Fixed(includeInToString = false, build = true) DoubleSortingScratch scratch
+    ) {
+        processDoubles(builder, position, values, percentile, scratch);
+    }
+
+    @Evaluator(extraName = "DoubleLong", warnExceptions = IllegalArgumentException.class)
+    static void process(
+        DoubleBlock.Builder builder,
+        int position,
+        DoubleBlock values,
+        long percentile,
+        @Fixed(includeInToString = false, build = true) DoubleSortingScratch scratch
+    ) {
+        processDoubles(builder, position, values, percentile, scratch);
+    }
+
+    @Evaluator(extraName = "DoubleDouble", warnExceptions = IllegalArgumentException.class)
+    static void process(
+        DoubleBlock.Builder builder,
+        int position,
+        DoubleBlock values,
+        double percentile,
+        @Fixed(includeInToString = false, build = true) DoubleSortingScratch scratch
+    ) {
+        processDoubles(builder, position, values, percentile, scratch);
+    }
+
+    @Evaluator(extraName = "IntegerInteger", warnExceptions = IllegalArgumentException.class)
+    static void process(
+        IntBlock.Builder builder,
+        int position,
+        IntBlock values,
+        int percentile,
+        @Fixed(includeInToString = false, build = true) IntSortingScratch scratch
+    ) {
+        processInts(builder, position, values, percentile, scratch);
+    }
+
+    @Evaluator(extraName = "IntegerLong", warnExceptions = IllegalArgumentException.class)
+    static void process(
```

Comment:
I think i might have used `Cast.cast` to force `percentile` to always be a `double`. I'm not sure it's worth having an evaluator for each one.

Reply:
Oh, I didn't know we had such magic. Checking it! It may simplify things a bit. At least remove a layer of functions 👀